apps/andi/lib/andi/input_schemas/datasets.ex

```diff
@@ -122,21 +128,72 @@ defmodule Andi.InputSchemas.Datasets do
       |> Changeset.get_field(:owner_id, nil)
       |> extract_owner_id(form_changes)
 
+    organization_id =
+      changeset
+      |> Changeset.get_field(:organization_id, nil)
+      |> extract_organization_id(form_changes)
+
     dataset_link =
       changeset
       |> Changeset.apply_changes()
       |> StructTools.to_map()
       |> Map.merge(form_changes)
       |> Map.get(:datasetLink)
 
-    case owner_id do
-      nil ->
-        existing_dataset |> update(%{technical: technical_changes, business: business_changes, id: dataset_id, datasetLink: dataset_link})
+    changes = %{
+      dataset_id: dataset_id,
+      technical_changes: technical_changes,
+      business_changes: business_changes,
+      owner_id: owner_id,
+      organization_id: organization_id,
+      dataset_link: dataset_link
+    }
 
-      owner_id ->
-        existing_dataset
-        |> update(%{technical: technical_changes, business: business_changes, id: dataset_id, owner_id: owner_id, datasetLink: dataset_link})
-    end
+    existing_dataset |> update_dataset(owner_id, org_changed, changes)
+  end
+
+  def update_dataset(existing_dataset, nil, false, changes) do
```

Comment:
Good use of function hydras. Is there a good way to make each of these heads smaller? I see some commonality between them, but I also don't want to make unnecessary repeat database calls for cleaner code.

Reply:
Wouldn't only one of these functions call with each underlying data update due to pattern matching?

Reply:
@ksmith-accenture - true!

@LtChae - we tried doing something like this:
`
 def update_from_form(dataset_id, form_changes) do
    org_changed = is_org_changed?(form_changes)
    existing_dataset = get(dataset_id)
    changeset = InputConverter.andi_dataset_to_full_ui_changeset(existing_dataset)

    technical_changes =
      changeset
      |> Changeset.get_change(:technical)
      |> Changeset.apply_changes()
      |> StructTools.to_map()
      |> Map.merge(form_changes)

    business_changes =
      changeset
      |> Changeset.get_change(:business)
      |> Changeset.apply_changes()
      |> StructTools.to_map()
      |> Map.merge(form_changes)

    owner_id =
      changeset
      |> Changeset.get_field(:owner_id, nil)
      |> extract_owner_id(form_changes)

    organization_id =
      changeset
      |> Changeset.get_field(:organization_id, nil)
      |> extract_organization_id(form_changes)

    dataset_link =
      changeset
      |> Changeset.apply_changes()
      |> StructTools.to_map()
      |> Map.merge(form_changes)
      |> Map.get(:datasetLink)

    updated_if_different_changes = %{
      owner_id: owner_id,
      organization_id: organization_id
    }

    always_updated_changes = %{
      dataset_id: dataset_id,
      technical_changes: technical_changes,
      business_changes: business_changes,
      dataset_link: dataset_link
    }

    existing_dataset |> update_dataset(owner_id, org_changed, always_updated_changes, updated_if_different_changes)
  end

  def update_dataset(existing_dataset, nil, false, always_updated_changes, _updated_if_different_changes) do
    existing_dataset
    |> update(always_updated_changes)
  end

  def update_dataset(existing_dataset, owner_id, false, always_updated_changes, updated_if_different_changes) do
    changes = Map.put(always_updated_changes, :owner_id, updated_if_different_changes.owner_id)
    existing_dataset
    |> update(changes)
  end

  def update_dataset(existing_dataset, nil, true, always_updated_changes, updated_if_different_changes) do
    changes = Map.put(always_updated_changes, :organization_id, updated_if_different_changes.organization_id)
    existing_dataset
    |> update(changes)
  end

  def update_dataset(existing_dataset, owner_id, true, always_updated_changes, updated_if_different_changes) do
    changes = Map.put(always_updated_changes, :organization_id, updated_if_different_changes.organization_id)
    changes = Map.put(always_updated_changes, :owner_id, updated_if_different_changes.owner_id)
    existing_dataset
    |> update(changes)
  end
`

However, after we started heading down this road, we found that it wasn't saving that many lines of code, and wasn't drastically improving readability. What would your thoughts be on leaving this as is?

Reply:
Spoke with Tim- decided to leave this as is!